improvement(ldap authorization): adding support to LDAP

fgelcer opened the pull request:
only authorization right now.
it is the whole code needed, but still no tests
were yet set to run with LDAP, but there is now
a configurations yaml to add to any test to run
with the LDAP.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

fruch commented on sdcm/cluster.py:
```diff
@@ -2789,8 +2845,12 @@ def terminate_node(self, node):
     def get_db_auth(self):
-        user = self.params.get('authenticator_user', default=None)
-        password = self.params.get('authenticator_password', default=None)
+        if self.params.get('use_ldap_authorization', default=None) and self.ldap_started:
```
please remove the `default` param, None is the default if flag wasn't defined

fgelcer replied:
ok

fruch commented on sdcm/cluster.py:
```diff
@@ -121,6 +122,12 @@ class Setup:
     BACKTRACE_DECODING = False
     INTRA_NODE_COMM_PUBLIC = False
     RSYSLOG_ADDRESS = None
+    LDAP_ADDRESS = None
```
why do need all those on the Setup class ? they are going to change during the test ?

fgelcer replied:
they will be populated once the container is created, and will be used across the test in different places...
the `LDAP_ADDRESS` will be used for create a connection to the LDAP server for example...

fruch replied:
the `LDAP_ADDRESS` o.k. but the rest are hardcodede values for now, like username/password and such. I don't see the reason for them to be on Setup class.

in some places you are using `LDAP_BASE_OBJECT` directly, and in some reading it from ` node.ldap_config['ldap_base_object']` and in some from `Setup.LDAP_BASE_OBJECT`

I would prefer it would be the first one only, until prove it's need to be dynamic or change during the run.
i.e. all those are known in advance, no need to put them all over the place.

fgelcer replied:
the only access to `node.ldap_config['ldap_base_object']` is when we set the `Setup.LDAP_BASE_OBJECT`, so its usage is pretty consistent

fruch replied:
you are missing the point, having one `LDAP_BASE_OBJECT` that all the code point to it, if it's a const value.
no need for all the wrapper around it.

`Setup` should be used for global values we are changing during he span of test, i.e. like addresses and ports.

Please remove any consts values out of `Setup` class.

fruch commented on sdcm/localhost.py:
```diff
@@ -33,6 +35,18 @@ def __init__(self, user_prefix: Optional[str] = None, test_id: Optional[str] = N
+                'ldap_ssl_port': ContainerManager.get_container_port(self, "ldap", LDAP_SSL_PORT)}
+
+    @property
+    def ldap_config(self) -> Optional[dict]:
```
how this configuration is related to the localhost class ? i'm not sure I'm following...

fgelcer replied:
the localhost configuration runs on the runner, and these configurations (or most of them) are sent to the container creation parameters...

fruch commented on sdcm/cluster.py:
```diff
@@ -2686,6 +2741,7 @@ def __init__(self, cluster_uuid=None, cluster_prefix='cluster', node_prefix='nod
         else:
             raise ValueError('Unsupported type: {}'.format(type(n_nodes)))
         self.coredumps = dict()
+        self.ldap_started = False
```
maybe `ldap_docker_running` ?

fgelcer replied:
ok

roydahan commented:
@scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
The first one should win (this) and @juliayakovlev will need to include this one on top of her change.

fruch commented:
> @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.

the first to be merged should win :smile: 
the looser get to rebuild his docker image after he rebase

fruch commented:
please rebase, and update the version in `docker/env/version` again and build the docker again.
next time please give a temporary name when building a new docker, i.e. `0.78_add_ldap`, and when getting close to approval, can be given the actual version and rebuild it.

fruch commented:
> > @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> > The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.
> 
> the first to be merged should win 
> the looser get to rebuild his docker image after he rebase

@fgelcer BTW you lost to @juliayakovlev  :stuck_out_tongue:

fgelcer commented:
> > > @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> > > The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.
> > 
> > 
> > the first to be merged should win
> > the looser get to rebuild his docker image after he rebase
> 
> @fgelcer BTW you lost to @juliayakovlev

it was intentional, as ladies first :-)

bentsi commented on docker/env/version:
```diff
@@ -1 +1 @@
-0.78
+0.80
```
You broke Hydra version 0.80, this is not the way we update the Hydra Docker image. Did you talk with somebody before doing this?

fruch replied:
also version `0.79` was burnt...
anyhow next time I think it would be clear, when building a version docker for testing a PR,
it should be `0.80_[some suffix to describe it]`, it should be rebuild into a formal version only once the PR is approved.
(which isn't the case yet)

fgelcer replied:
> You broke Hydra version 0.80, this is not the way we update the Hydra Docker image. Did you talk with somebody before doing this?

what was broken with v0.80?

yes, I spoke the guys, and as @fruch  mentioned in his comment [here](https://github.com/scylladb/scylla-cluster-tests/pull/2761#issuecomment-704538833), next time, i will create it once getting close to approval (as I expected it to be) or once approved (and then change the version file + build and push the image)...

fruch reviewed (changes requested):
static vaule should go out of `Setup` class
* LDAP_USERNAME
* LDAP_PASSWORD
* LDAP_BASE_OBJECT
* LDAP_USERS
* LDAP_ROLE

fgelcer commented:
> static vaule should go out of `Setup` class
> 
> * LDAP_USERNAME
> * LDAP_PASSWORD
> * LDAP_BASE_OBJECT
> * LDAP_USERS
> * LDAP_ROLE

why is the `LDAP_ADDRESS` different from these other parameters?

fruch commented:
@fgelcer this would need rebuilding of the docker cause of #2789

roydahan commented:
@bentsi / @fruch Except of the docker version issue that @fgelcer will handle right before merging (so it won't rotten).
Any other feedback?
If not, please mark as approved.

fruch commented on sdcm/localhost.py:
```diff
@@ -33,6 +35,18 @@ def __init__(self, user_prefix: Optional[str] = None, test_id: Optional[str] = N
+                'ldap_ssl_port': ContainerManager.get_container_port(self, "ldap", LDAP_SSL_PORT)}
+
+    @property
+    def ldap_config(self) -> Optional[dict]:
```
this property can be remove, it's not used anywhere

fruch reviewed (changes requested):
`localhost.ldap_config` property can be removed, since it's not used. (and need to fix a conflict on this file anyhow)
beside that, looks o.k.

fgelcer commented:
> `localhost.ldap_config` property can be removed, since it's not used. (and need to fix a conflict on this file anyhow)
> beside that, looks o.k.

everything was updated...
just, once approved, and before it will be merged, the `docker/env/version` need to be updated to the final value of it

fgelcer commented:
conflicts were resolved, and tests are making very good progress.
ping @fruch @roydahan

roydahan commented on configurations/ldap-authorization.yaml:
```diff
@@ -0,0 +1,5 @@
+authenticator: 'PasswordAuthenticator'
+authenticator_user: cassandra
+authenticator_password: cassandra
+authorizer: 'CassandraAuthorizer'
```
Is this what required when using LDAP?

fgelcer replied:
yes, as this is the "authorizer"

roydahan commented on configurations/ldap-authorization.yaml:
```diff
@@ -0,0 +1,5 @@
+authenticator: 'PasswordAuthenticator'
+authenticator_user: cassandra
+authenticator_password: cassandra
```
Why are these relevant to this configuration?

fgelcer replied:
as we are setting the `authorizer`, and if the user and password weren't set up before, then it need to be set up in here altogether, otherwise, it won't work

roydahan replied:
But this is the authenticator, not autorizer.

fgelcer replied:
this is what will allow one user to enter a username and password

roydahan commented on sdcm/cluster_aws.py:
```diff
@@ -867,6 +867,7 @@ def node_config_setup(self, node, seed_address=None, endpoint_snitch=None, murmu
             alternator_enforce_authorization=self.params.get('alternator_enforce_authorization'),
             internode_compression=self.params.get('internode_compression'),
             internode_encryption=self.params.get('internode_encryption'),
+            ldap=self.params.get('use_ldap_authorization', False),
```
the default should be in sct_config

fgelcer replied:
how do you set this on sct_config?

roydahan commented on sdcm/cluster_k8s/__init__.py:
```diff
@@ -583,7 +583,8 @@ def node_config_setup(self,
                           alternator_enforce_authorization=self.params.get("alternator_enforce_authorization"),
-                          internode_compression=self.params.get("internode_compression"))
+                          internode_compression=self.params.get("internode_compression"),
+                          ldap=self.params.get('use_ldap_authorization', False))
```
same

fgelcer replied:
removed

roydahan reviewed (commented):
Overall looks good to me.
Had some minor comments you need to address.

In general, you're using too much positional argument when calling functions, which is bad practice for future additions and refactors.
I prefer you would use keyword arguments like: ip=ip, port=port, etc.

fgelcer commented:
@roydahan / @fruch ping

fgelcer commented:
> @roydahan / @fruch ping

rebased on latest master, and already resolved the conflicts there were with this PR

fgelcer commented:
@fruch @roydahan , i see that someone created a new `docker/env/version` so once this one is approved (and ready to merge) i will re-create another docker image and push it to dockerhub

fgelcer commented:
@roydahan , fixed conflicts (changed the docker version) and added the 2 pipelines we spoke of. I will now, while this isn't yet approved and merged, create the jobs (for enterprise and 2020.1) and will give them a ride.
@fruch , ready to merge (as you have already reviewed it)

fgelcer commented:
@fruch / @roydahan , i have created the jobs for 2020.1 and enterprise, and i ran the 2020.1 job (short longevity with LDAP enabled) and it passed:
https://jenkins.scylladb.com/view/enterprise-2020.1/job/enterprise-2020.1/job/SCT_Enterprise_Features/job/LDAP_authorization/job/longevity-100gb-4h-with-ldap-authorization/

fgelcer commented:
@fruch ping

fruch reviewed (approved):
LGTM

fruch commented:
Fabio, please upload the formal docker, so it can be merged

fgelcer commented:
> Fabio, please upload the formal docker, so it can be merged

@fruch , done

fgelcer commented:
@scylladb/qa-maintainers ping

fruch reviewed (approved):
LGTM
